static/js/narrow.js

```diff
@@ -272,6 +258,61 @@ export function activate(raw_operators, opts) {
         id_info.target_id = Number.parseInt(filter.operands("id")[0], 10);
     }
 
+    // Narrow with near / id filer. Match stream and topic with that of the
+    // message in case the message was moved after the link was created. This
+    // ensures near / id links work as permalinks, i.e. don't break even after
+    // the message was moved.
+    if (id_info.target_id && filter.has_operator("stream")) {
+        const target_message = message_store.get(id_info.target_id);
+
+        function update_raw_operators(operators, message) {
+            for (const index in operators) {
+                if (operators[index].operator === "stream") {
+                    operators[index].operand = message.display_recipient;
+                } else if (operators[index].operator === "topic") {
+                    operators[index].operand = message.topic;
+                }
+            }
+            return operators;
+        }
+
+        // Fetch message from database, if it is not available locally.
+        if (!target_message) {
+            channel.get({
+                url: `/json/messages/${id_info.target_id}`,
+                idempotent: true,
+                success(data) {
+                    // After the message is fetched, we make the message locally
+                    // available and try to re-active the narrow with the same filters.
+                    message_helper.process_new_message(data.message);
+                    activate(raw_operators, opts);
+                },
+            });
+            // No need to active this narrow as it will be handled by
+            // the success handler above.
+            return;
```

Comment:
I think this logic needs a handler for what happens if the target message ID doesn't exist; it seems like right now it'll fail if e.g. the message was deleted.